[sdpa] make sure to recompile if alignment is different than before

ColinPeppler opened the pull request:
## Context
An example from Qwen2-7B
- This come from running torch.compile with a sequence length that is
divisible by 8 (no padding needed). Call this `Run1`.
- If we then run the compiled model with a difference length that isn't
divisible by 8 (requires padding). Call this `Run2`.
- Then we'll see this error.
```
File "/var/tmp/torchinductor_nobody/2w/c2wby7ilxbna45xrtrrfjqpeutwouruviu2742ockunnd2bleeiz.py", line 1963, in call
    buf24 = torch.ops.aten._scaled_dot_product_efficient_attention_backward.default(reinterpret_tensor(buf18, (s85, 3584 // s19, s48, 512 // (512 // s19)), (s48*(512 // (512 // s19))*(3584 // s19), 512 // (512 // s19), (512 // (512 // s19))*(3584 // s19), 1), 0), buf20, buf21, buf22, buf23, getitem, getitem_1, getitem_2, getitem_3, 0.0, [True, True, True, False], scale=0.08838834764831845)
File "torch/_ops.py", line 841, in __call__
    return self._op(*args, **kwargs)
RuntimeError: attn_bias is not correctly aligned (strideM). attn_bias.stride(2) = 6102, and should be a multiple of 4.
```
- We only see the error because we did not recompile on `Run2`. Instead we ran the inputs on the same graph as `Run1`.


### A bit more on why.
Here we check whether to realize the unpadded buffer (unwrapped slice) which we want for `Run1` but not for `Run2`.
https://github.com/pytorch/pytorch/blob/0897affcd5e0e5281db7202408ca637bbf2f7abe/torch/_inductor/lowering.py#L2687-L2694

## Fix
Size hint doesn't guard, so the fix is to use `guard_or*` to guard. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163083



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

ColinPeppler committed:
[sdpa] make sure to recompile if alignment is different than before

[ghstack-poisoned]

ColinPeppler commented on torch/_inductor/ir.py:
```diff
@@ -446,14 +446,13 @@ def is_aligned_realized_tensor_hint(
 
     aligned_strides = all(
-        (V.graph.sizevars.size_hint_or_throw(x.get_stride()[i]) % alignment) == 0
+        (V.graph.sizevars.guard_or_false(sympy.Eq(x.get_stride()[i] % alignment, 0)))
```
@laithsakka a bit concerned that we might be overguarding with this

Would a better approach be to keep the sizehint condition? So, if the sizehint condition is True, then we can guard. If the sizehint conidition is false we don't need to guard.

eellison replied:
The other alternative is 

if statically_known_true:
- fast path
else:
- copy

I think this might be better. Wdyt @laithsakka ?

laithsakka replied:
we can use statically known true, 
or you can create a single big sympy expressioons that represent the condition and call guard_or_false once this way you wont be over guarding. 

cond = sympyTrue
for i in range(len(x.get_stride()) - 1):
  cond = sympy.And(cond, sympy.Eq(x.get_stride()[i] % alignment, 0))
  
  then do 
  if V.graph.sizevars.guard_or_false(cond)
  
  Note also please instead of using % use   Mod class in 
  /home/lsakka/pytorch/torch/utils/_sympy/functions.py

ColinPeppler replied:
okay the long sympy expr isn't too bad

```
    // A. check then guard
    aligned_strides = all(
        V.graph.sizevars.statically_known_multiple_of(x.get_stride()[i], alignment)
        for i in range(len(x.get_stride()) - 1)
    )
    aligned_last_dim = (
        V.graph.sizevars.statically_known_equals(x.get_size()[-1], 1)
        or V.graph.sizevars.statically_known_leq(x.get_size()[-1], 1)
    )
    if aligned_strides and aligned_last_dim:
        # make sure to guard
        guard_aligned_strides = sympy.And(
            *(sympy.Eq(Mod(s, alignment), 0) 
            for s in x.get_stride()[:-1])
        )
        guard_aligned_last_dim = sympy.Or(
            sympy.Eq(x.get_stride()[-1], 1),
            sympy.Le(x.get_size()[-1], 1)
        )
        V.graph.sizevars.expect_true(guard_aligned_strides)
        V.graph.sizevars.expect_true(guard_aligned_last_dim)
        return True
    return False
    
    // B. long sympy expr
    aligned_strides = sympy.And(
        *(sympy.Eq(Mod(s, alignment), 0) 
        for s in x.get_stride()[:-1])
    )
    aligned_last_dim = sympy.Or(
        sympy.Eq(x.get_stride()[-1], 1),
        sympy.Le(x.get_size()[-1], 1)
    )
    is_aligned = sympy.And(aligned_strides, aligned_last_dim)
    return V.graph.sizevars.guard_or_false(is_aligned)
```

ColinPeppler committed:
Update on "[sdpa] make sure to recompile if alignment is different than before"

An example from Qwen2-7B
- This come from running torch.compile with a sequence length that is
divisible by 8 (no padding needed). Run1.
- If we then run the compiled model with a difference length that isn't
divisible by 8 (requires padding). Run2.
- Then we'll see this error.
> RuntimeError: attn_bias.stride(2) = 6102, and should be a multiple of 4.
- We only see the error because we did not recompile on Run 2.

Size hint doesn't guard, so the fix is to use `guard_or*` to guard. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ColinPeppler committed:
Update on "[sdpa] make sure to recompile if alignment is different than before"

An example from Qwen2-7B
- This come from running torch.compile with a sequence length that is
divisible by 8 (no padding needed). Run1.
- If we then run the compiled model with a difference length that isn't
divisible by 8 (requires padding). Run2.
- Then we'll see this error.
> RuntimeError: attn_bias.stride(2) = 6102, and should be a multiple of 4.
- We only see the error because we did not recompile on Run 2.

Size hint doesn't guard, so the fix is to use `guard_or*` to guard. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ColinPeppler committed:
Update on "[sdpa] make sure to recompile if alignment is different than before"

## Context
An example from Qwen2-7B
- This come from running torch.compile with a sequence length that is
divisible by 8 (no padding needed). Call this `Run1`.
- If we then run the compiled model with a difference length that isn't
divisible by 8 (requires padding). Call this `Run2`.
- Then we'll see this error.
```
File "/var/tmp/torchinductor_nobody/2w/c2wby7ilxbna45xrtrrfjqpeutwouruviu2742ockunnd2bleeiz.py", line 1963, in call
    buf24 = torch.ops.aten._scaled_dot_product_efficient_attention_backward.default(reinterpret_tensor(buf18, (s85, 3584 // s19, s48, 512 // (512 // s19)), (s48*(512 // (512 // s19))*(3584 // s19), 512 // (512 // s19), (512 // (512 // s19))*(3584 // s19), 1), 0), buf20, buf21, buf22, buf23, getitem, getitem_1, getitem_2, getitem_3, 0.0, [True, True, True, False], scale=0.08838834764831845)
File "torch/_ops.py", line 841, in __call__
    return self._op(*args, **kwargs)
RuntimeError: attn_bias is not correctly aligned (strideM). attn_bias.stride(2) = 6102, and should be a multiple of 4.
```
- We only see the error because we did not recompile on `Run2`. Instead we ran the inputs on the same graph as `Run1`.


### A bit more on why.
Here we check whether to realize the unpadded buffer (unwrapped slice) which we want for `Run1` but not for `Run2`.
https://github.com/pytorch/pytorch/blob/0897affcd5e0e5281db7202408ca637bbf2f7abe/torch/_inductor/lowering.py#L2687-L2694

## Fix
Size hint doesn't guard, so the fix is to use `guard_or*` to guard. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

eellison reviewed (approved):
Looks good, thanks ! Nice test case as well

ColinPeppler commented:
@pytorchbot merge

ColinPeppler committed:
Update on "[sdpa] make sure to recompile if alignment is different than before"

## Context
An example from Qwen2-7B
- This come from running torch.compile with a sequence length that is
divisible by 8 (no padding needed). Call this `Run1`.
- If we then run the compiled model with a difference length that isn't
divisible by 8 (requires padding). Call this `Run2`.
- Then we'll see this error.
```
File "/var/tmp/torchinductor_nobody/2w/c2wby7ilxbna45xrtrrfjqpeutwouruviu2742ockunnd2bleeiz.py", line 1963, in call
    buf24 = torch.ops.aten._scaled_dot_product_efficient_attention_backward.default(reinterpret_tensor(buf18, (s85, 3584 // s19, s48, 512 // (512 // s19)), (s48*(512 // (512 // s19))*(3584 // s19), 512 // (512 // s19), (512 // (512 // s19))*(3584 // s19), 1), 0), buf20, buf21, buf22, buf23, getitem, getitem_1, getitem_2, getitem_3, 0.0, [True, True, True, False], scale=0.08838834764831845)
File "torch/_ops.py", line 841, in __call__
    return self._op(*args, **kwargs)
RuntimeError: attn_bias is not correctly aligned (strideM). attn_bias.stride(2) = 6102, and should be a multiple of 4.
```
- We only see the error because we did not recompile on `Run2`. Instead we ran the inputs on the same graph as `Run1`.


### A bit more on why.
Here we check whether to realize the unpadded buffer (unwrapped slice) which we want for `Run1` but not for `Run2`.
https://github.com/pytorch/pytorch/blob/0897affcd5e0e5281db7202408ca637bbf2f7abe/torch/_inductor/lowering.py#L2687-L2694

## Fix
Size hint doesn't guard, so the fix is to use `guard_or*` to guard. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ColinPeppler commented on test/inductor/test_cuda_repro.py:
```diff
@@ -1348,7 +1348,7 @@ def forward(self, x):
             out3 = model(input3)
 
-        self.assertEqual(cnts.frame_count, 1)
+        self.assertEqual(cnts.frame_count, 2)
```
@eellison looks like this one didn't need a recompile on S=513 for flash attention, 

not sure if I should have separate set of guards for flash attn vs. memory efficient attn vs. cudnn...
